task/task_service.go

```diff
@@ -0,0 +1,190 @@
+	// object is actually the refreshed version of the provided one.
+	MarkAsFailedWithError(taskObject *TaskObject, err error) (*TaskObject, error)
+
+	// PersistState writes the given task object to the configured Backend.
+	PersistState(taskObject *TaskObject) error
+
+	// WaitForFinalStatus blocks and waits for the given task to reach a final
+	// status. The given closer can end the waiting and thus stop blocking the
+	// call to WaitForFinalStatus. The returned task object is actually the
+	// refreshed version of the provided one.
+	WaitForFinalStatus(taskObject *TaskObject, closer <-chan struct{}) (*TaskObject, error)
+}
+
+// TaskServiceConfig represents the configurations for the task service that is
+// going to be created.
+type TaskServiceConfig struct {
+	Backend Backend
+}
+
+// DefaultTaskServiceConfig returns a best effort default configuration for the
+// task service.
+func DefaultTaskServiceConfig() TaskServiceConfig {
+	newConfig := TaskServiceConfig{
+		Backend: NewMemoryBackend(),
+	}
+
+	return newConfig
+}
+
+// NewTaskService returns a new configured task service instance.
+func NewTaskService(config TaskServiceConfig) TaskService {
+	newTaskService := &taskService{
+		TaskServiceConfig: config,
+	}
+
+	return newTaskService
+}
+
+type taskService struct {
+	TaskServiceConfig
+}
+
+func (ts *taskService) Create(action Action) (*TaskObject, error) {
+	taskObject := &TaskObject{
+		ID:           uuid.NewV4().String(),
+		ActiveStatus: StatusStarted,
+		FinalStatus:  "",
+	}
+
+	go func() {
+		err := action()
+		if err != nil {
+			_, markErr := ts.MarkAsFailedWithError(taskObject, err)
+			if markErr != nil {
+				fmt.Printf("[E] Task.MarkAsFailed failed: %#v\n", maskAny(markErr))
+				return
+			}
+			return
+		}
+
+		_, err = ts.MarkAsSucceeded(taskObject)
+		if err != nil {
+			fmt.Printf("[E] Task.MarkAsSucceeded failed: %#v\n", maskAny(err))
+			return
+		}
+	}()
+
+	err := ts.PersistState(taskObject)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+
+	return taskObject, nil
+}
+
+func (ts *taskService) FetchState(taskObject *TaskObject) (*TaskObject, error) {
+	var err error
+
+	taskObject, err = ts.Backend.Get(taskObject.ID)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+
+	return taskObject, nil
+}
+
+func (ts *taskService) MarkAsFailedWithError(taskObject *TaskObject, err error) (*TaskObject, error) {
+	taskObject.ActiveStatus = StatusStopped
+	taskObject.Error = err.Error()
+	taskObject.FinalStatus = StatusFailed
+
+	err = ts.PersistState(taskObject)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+
+	return taskObject, nil
+}
+
+func (ts *taskService) MarkAsSucceeded(taskObject *TaskObject) (*TaskObject, error) {
+	taskObject.ActiveStatus = StatusStopped
+	taskObject.FinalStatus = StatusSucceeded
+
+	err := ts.PersistState(taskObject)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+
+	return taskObject, nil
+}
+
+func (ts *taskService) PersistState(taskObject *TaskObject) error {
```

Comment:
No. My intention was to define an interface that defines the interface of a implementation. So the interface provides all building blocks to make the service work. The interface here does not care about visibility, but logic. Persisting state is a logical component that needs to be fulfilled. Thus the interface requires that logic to be implemented.